src/client/automation/playback/type/type-text.js

```diff
@@ -84,26 +105,68 @@ function _excludeInvisibleSymbolsFromSelection (selection) {
     return selection;
 }
 
+// NOTE: Typing can be prevented in Chrome/Edge but can not be prevented in IE11 or Firefox
+//       Firefox does not support TextInput event
+//       Safari supports the TextInput event but has a bug: e.data is added to the node value.
+//       So in Safari we need to call preventDefault in the last textInput handler but not prevent the Input event
+
+let forceInputInSafari;
+
+function simulateTextInput (element, text) {
```

Comment:
In this method almost all `textInput` logic is written for Safari. May be it would be more clear if we e.g.: 

```js
function simulateTextInput (element, text) {
    //logic only for Safari
}

var beforeContentChanged = () => {
        if (browserUtils.isSafari)
            needProcessInput = simulateTextInput(element, text);
        else
            needProcessInput = browserUtils.isFirefox || eventSimulator.textInput(element, text) || browserUtils.isIE11;

        needRaiseInputEvent = needProcessInput && !browserUtils.isIE11;
};	
```
What do you think @AlexKamaev @AndreyBelym ?

Reply:
we use simulateTextInput not only if `beforeContentChanged` but also in `_typeTextToTextEditable`.
If we detach safari logic from `eventSimulator.textInput(...)` we will need to duplicate safari logic

Reply:
Oh, ok then